evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -1,34 +1,147 @@
+from collections import (
+    defaultdict,
+)
+
 import logging
 
+from cytoolz import (
+    pipe,
+)
+
+import rlp
+
 from web3.contract import (
     Contract,
 )
 
 from eth_utils import (
+    event_signature_to_log_topic,
     is_canonical_address,
     to_checksum_address,
     to_dict,
+    to_tuple,
+)
+
+from evm.rlp.sedes import (
+    address,
+    hash32,
 )
 
 from evm.utils.hexadecimal import (
+    encode_hex,
     decode_hex,
 )
+from evm.utils.numeric import (
+    big_endian_to_int,
+)
 
 from evm.vm.forks.sharding.config import (
     get_sharding_config,
 )
 
 
+class NextLogUnavailable(Exception):
+    pass
+
+
+@to_dict
+def parse_collation_added_data(data_hex):
+    data_bytes = decode_hex(data_hex)
+    score = big_endian_to_int(data_bytes[-32:])
+    is_new_head = bool(big_endian_to_int(data_bytes[-64:-32]))
+    header_bytes = data_bytes[:-64]
+    # [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, num, bytes]
+    sedes = rlp.sedes.List([
+        rlp.sedes.big_endian_int,
+        rlp.sedes.big_endian_int,
+        hash32,
+        hash32,
+        hash32,
+        address,
+        hash32,
+        hash32,
+        rlp.sedes.big_endian_int,
+        rlp.sedes.binary,
+    ])
+    header_values = rlp.decode(header_bytes, sedes=sedes)
+    yield 'header', header_values
+    yield 'is_new_head', is_new_head
+    yield 'score', score
+
+
 class VMC(Contract):
 
-    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
+    logger = logging.getLogger("evm.chain.sharding.VMC")
+
+    # For handling logs filtering
+    # Event:
+    #   CollationAdded(indexed uint256 shard, bytes collationHeader, bool isNewHead, uint256 score)
+    COLLATION_ADDED_TOPIC = event_signature_to_log_topic(
+        "CollationAdded(int128,bytes4096,bool,int128)"
+    )
 
-    def __init__(self, *args, default_privkey, **kwargs):
+    def __init__(self, *args, log_handler, default_privkey, **kwargs):
+        self.log_handler = log_handler
```

Comment:
I think you're going to need one log handler per shard.  Since the log handler is stateful, each time you ask for more logs, it only returns the ones for the given `shard_id`.  If you're tracking multiple `shard_id`s then I think you're going to miss logs.

Reply:
Ahh I didn't thought of this case. Thank you for discovering that.
Will fix it.

Reply:
Actually I was thinking that we could use maintain one `recent_block_hashes` for all shards, to share it over shards. But in this way we will need to store all logs from all shards. It seems not better than keeping one log_handler for each shard.

Reply:
Yeah, one handler per shard seems cleanest despite the added overhead.  We can address the performance issues later if needed.